cirq/ops/controlled_operation_test.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import pytest
+
+import numpy as np
+import cirq
+from cirq import protocols
+
+
+def test_controlled_operation_init():
+    cb = cirq.NamedQubit('ctr')
+    q = cirq.NamedQubit('q')
+    g = cirq.Gate()
+    v = cirq.GateOperation(g, (q,))
+    c = cirq.ControlledOperation(cb, v)
+    assert c.sub_operation == v
+    assert c.control == cb
+    assert c.qubits == (cb, q)
+
+    eq = cirq.testing.EqualsTester()
+    eq.add_equality_group(c, c.with_qubits(cb, q))
+
+
+def test_controlled_operation_eq():
+    c1 = cirq.NamedQubit('c1')
+    q1 = cirq.NamedQubit('q1')
+    c2 = cirq.NamedQubit('c2')
+
+    eq = cirq.testing.EqualsTester()
+
+    eq.make_equality_group(lambda: cirq.ControlledOperation(c1, cirq.X(q1)))
+    eq.make_equality_group(lambda: cirq.ControlledOperation(c2, cirq.X(q1)))
+    eq.make_equality_group(lambda: cirq.ControlledOperation(c1, cirq.Z(q1)))
+    eq.make_equality_group(lambda: cirq.ControlledOperation(c2, cirq.Z(q1)))
+
+
+def test_str():
+    c1 = cirq.NamedQubit('c1')
+    c2 = cirq.NamedQubit('c2')
+    q2 = cirq.NamedQubit('q2')
+
+    assert (str(cirq.ControlledOperation(c1, cirq.CZ(c2, q2))) ==
+            "C(c1)CZ(c2, q2)")
```

Comment:
Should be `"CCZ(c1, c2, q2)"`

Reply:
Oh, I see now that this might be difficult. The current str is acceptable.

Reply:
We can't assume that.

Reply:
if we can't assume anything about the structure of the suboperation's `str`, then this is our best bet.